src/data/mv_polynomial/supported.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2021 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import ring_theory.adjoin.basic
+import data.mv_polynomial.variables
+import linear_algebra.finsupp
+/-!
+# Polynomials supported by a set of variables
+
+This file contains the definition and lemmas about `mv_polynomial.supported`.
+
+## Main definitions
+
+* `mv_polynomial.supported` : Given a set `s : set σ`, `supported R s` is the subalgebra of
+  `mv_polynomial σ R` consisting of polynomials whose set of variables is contained in `s`.
+
+## Tags
+variables, polynomial, vars
+-/
+universes u v w
+
+namespace mv_polynomial
+variables {σ τ : Type*} {R : Type u} {S : Type v} {r : R} {e : ℕ} {n m : σ}
+
+section comm_semiring
+variables [comm_semiring R] {p q : mv_polynomial σ R}
+
+variables (R)
+
+/-- The set of polynomials whose variables are contained in `s` as a `subalgebra` over `R`. -/
+noncomputable def supported (s : set σ) : subalgebra R (mv_polynomial σ R) :=
+algebra.adjoin R (X '' s)
+
+variables {σ R}
+
+open_locale classical
+open algebra
+
+lemma supported_eq_range_rename (s : set σ) :
+  supported R s = (rename (coe : s → σ)).range :=
+by rw [supported, set.image_eq_range, adjoin_range_eq_range_aeval, rename]
+
+variables {s t : set σ}
+
+lemma mem_supported : p ∈ (supported R s) ↔ ↑p.vars ⊆ s :=
+begin
+  rw [supported_eq_range_rename, alg_hom.mem_range],
+  split,
+  { rintros ⟨p, rfl⟩,
+    refine trans (finset.coe_subset.2 (vars_rename _ _)) _,
+    simp },
+  { intros hs,
+    exact exists_rename_eq_of_vars_subset_range p (coe : s → σ) subtype.val_injective (by simpa) }
+end
+
+lemma supported_eq_vars_subset : (supported R s : set (mv_polynomial σ R)) = {p | ↑p.vars ⊆ s} :=
+set.ext $ λ _, mem_supported
+
+@[simp] lemma mem_supported_vars (p : mv_polynomial σ R) : p ∈ supported R (↑p.vars : set σ) :=
+by rw [mem_supported]
+
+variable (s)
+
+lemma supported_eq_adjoin_X : supported R s = algebra.adjoin R (X '' s) := rfl
+
+@[simp] lemma supported_univ : supported R (set.univ : set σ) = ⊤ :=
+by simp [algebra.eq_top_iff, mem_supported]
+
+@[simp] lemma supported_empty : supported R (∅ : set σ) = ⊥ :=
+by simp [supported_eq_adjoin_X]
+
+@[simp] lemma supported_eq_top_of_subsingleton [subsingleton R] : supported R s = ⊤ :=
+begin
+  ext p,
+  haveI := (@mv_polynomial.unique R σ _),
+  rw [subsingleton.elim p 0],
+  simp [subalgebra.zero_mem],
+end
```

Comment:
Is there a more general statement hiding here that `subsingleton (subalgebra R A)` given `subsingleton R`?

Reply:
Yes, I guess that should be added.